Support formatting URL segments via new FORMAT_LINKS setting

platinumazure opened the pull request:
Fixes #790

## Description of the Change

Added new setting `JSON_API_FORMAT_LINKS`. This setting is used to inflect serializer properties (i.e., field names) on URLs specifically.

This allows users to have different inflections for properties in JSON responses (using `JSON_API_FORMAT_FIELD_NAMES`, existing behavior) vs properties in relationship and related resource links (using `JSON_API_FORMAT_LINKS`, new behavior, previously no inflection was supported).

I used `JSON_API_FORMAT_LINKS` as the setting name, but I feel it's very important we select a setting name that is as intuitive for users as possible. So, if we think this name could be improved, I'd love to discuss that here and then make a change once we have reached consensus.

This is my first contribution to a Python project, so please rip this apart :smile: Note that I was able to run `black`, `flake8`, and `pytest` individually, but I was not able to get `tox` to work, so I apologize in advance if I've missed something as a result.

## Checklist

- [x] PR only contains one change (considered splitting up PR)
- [x] unit-test added
- [x] documentation updated
- [x] `CHANGELOG.md` updated (only for user relevant changes)
- [x] author name in `AUTHORS`

sliverc commented on tests/test_views.py:
```diff
@@ -0,0 +1,51 @@
+def test_get_related_field_name_handles_formatted_link_segments(format_links):
+    url_segment = format_value(related_model_field_name, format_links)
+
+    request = RequestFactory().get("/basic_models/1/{}".format(url_segment))
```
use [RequestFactory fixture](https://pytest-django.readthedocs.io/en/latest/helpers.html#rf-requestfactory) of pytest-django so it is extendible if we need to in the future. Also a good stop to use [f-strings](https://realpython.com/python-f-strings/#simple-syntax)

sliverc commented on tests/test_views.py:
```diff
@@ -0,0 +1,51 @@
+related_model_field_name = "related_field_model"
+
+
+@pytest.mark.urls("tests.urls")
```
I think this line can be removed as you are using a request factory below and are not relying on urls in the test code.

sliverc commented on tests/urls.py:
```diff
@@ -0,0 +1,22 @@
+router = SimpleRouter()
+router.register(r"basic_models", BasicModelViewSet, basename="basic-model")
+
+urlpatterns = [
```
I am a bit on the fence with this as we are currently just migrating to the new test module and are not sure how things are turning out. But one issue in the past we had is that we had too much shared test code between different tests making maintanance tricky. As this is currently only needed in test_relations.py I prefer we move it into that module for now. We can always extract it again if need to later on.

platinumazure replied:
Your reasoning definitely makes sense. However, I don't know how to override the urlconf machinery besides using the decorator `@pytest.mark.urls()` and pointing to a separate module. Do you know how to tell Django to use a urlconf that is "inline" within the test module? If you can point me in the right direction, I'm happy to make this change.

sliverc replied:
The urlpatterns needs to be moved into `test_relations.py` top level and then you can use the following url marker:

```python
@pytest.mark.urls(__name__)
```

In [test_parsers.py](https://github.com/django-json-api/django-rest-framework-json-api/blob/d8f26fc9f87a8c684a80eafb10fe05524f3b3856/example/tests/test_parsers.py) we already do this but instead of using the urls marker it overwrites [ROOT_URLCONF](https://github.com/django-json-api/django-rest-framework-json-api/blob/d8f26fc9f87a8c684a80eafb10fe05524f3b3856/example/tests/test_parsers.py#L136) setting. I prefer the marker though.

platinumazure replied:
Perfect, I'll give that a shot. Thanks!

sliverc commented on tests/views.py:
```diff
@@ -0,0 +1,12 @@
+from .models import BasicModel
+
+
+class BasicModelViewSet(ModelViewSet):
```
Same as above better to be moved to test_relations.py for now.

platinumazure replied:
This effectively depends on figuring out how to "inline" the urlconf, as I asked about in the other unresolved conversation in this review. Once we can figure out how to move the urlconf inline, this change should be pretty trivial.

sliverc commented on tests/test_relations.py:
```diff
@@ -0,0 +1,46 @@
+    field.field_name = "relatedField_name"
+
+    expected = {
+        "self": "/basic_models/{}/relationships/{}/".format(
```
using [f-strings](https://realpython.com/python-f-strings/#simple-syntax) I think makes this a bit more readable.

sliverc reviewed (changes requested):
Thanks for your PR. Great work! There are some small comments I've added otherwise the PR is shaping up nicely.

Concerning naming of `FORMAT_LINKS`: I see that this could be misleading. But I can only come up with names which are really long and as the documentation states it well how it is used I think it is OK for me to leave it with this name (unless we come up with something better)

platinumazure commented:
Thanks @sliverc for the review!

I've addressed most of your feedback. The only outstanding item at this point is how to "inline" the urlconf in `tests/urls.py`; the URLs must be set up in order for the RelatedMixin links to reverse-resolve properly. So I need to figure out how to move them but still be able to use `@pytest.mark.urls`, or else I need guidance on mocking the reverse-resolution functionality to avoid depending on a urlconf.

If you have any suggestions off the top of your head, I would be grateful to hear them. Otherwise, I'll try to look into it some more in the next week or so.

platinumazure commented:
I've made the requested changes, and I believe this is ready for another review. Thanks!

sliverc reviewed (approved):
Thanks! LGTM

platinumazure commented:
Oh darn it, I forgot to change the name to what I had suggested in my other comment.

@sliverc Thoughts on `JSON_API_FORMAT_RELATED_FIELD_LINKS` or `JSON_API_FORMAT_RELATED_LINKS`? Slightly more specific but not too wordy IMO

sliverc commented:
I think in the future if there are other parts of the urls being automatically determined by DJA the setting should be called `FORMAT_LINKS`. Currently this PR only affects related links so renaming it to `FORMAT_RELATED_LINKS` is a good idea. If in the future if more parts of the various links are generated automatically we can always deprecate `FORMAT_RELATED_LINKS` and move towards using `FORMAT_LINKS`.

It would be great if you could create a PR. Thanks.

platinumazure commented:
Will do, I'll have another PR up in the next day or two.
